app/src/main/cpp/HandModels.cpp

```diff
@@ -0,0 +1,356 @@
+);
+
+const Material u_material = Material(
+  vec4(0.85, 0.85, 0.85, 1.0),
+  vec4(0.5, 0.5, 0.5, 1.0),
+  vec4(0.5, 0.5, 0.5, 1.0),
+  0.75f
+);
+
+vec4 calculate_light(vec4 norm, Light light, Material material) {
+  vec4 result = vec4(0.0f, 0.0f, 0.0f, 0.0f);
+  vec4 direction = -normalize(u_view * vec4(light.direction.xyz, 0.0f));
+  vec4 hvec;
+  float ndotl;
+  float ndoth;
+  result += light.ambient * material.ambient;
+  ndotl = max(0.0f, dot(norm, direction));
+  result += (ndotl * light.diffuse * material.diffuse);
+  hvec = normalize(direction + vec4(0.0f, 0.0f, 1.0f, 0.0f));
+  ndoth = dot(norm, hvec);
+  if (ndoth > 0.0f) {
+    result += (pow(ndoth, material.specularExponent) * material.specular * light.specular);
+  }
+  return result;
+}
+
+void main(void) {
+  vec4 pos = vec4(a_position, 1.0f);
+  vec4 localPos1 = u_jointMatrices[int(a_jointIndices.x)] * pos;
+  vec4 localPos2 = u_jointMatrices[int(a_jointIndices.y)] * pos;
+  vec4 localPos3 = u_jointMatrices[int(a_jointIndices.z)] * pos;
+  vec4 localPos4 = u_jointMatrices[int(a_jointIndices.w)] * pos;
+  vec4 localPos = localPos1 * a_jointWeights.x
+                + localPos2 * a_jointWeights.y
+                + localPos3 * a_jointWeights.z
+                + localPos4 * a_jointWeights.w;
+  gl_Position = u_perspective * u_view * u_model * localPos;
+
+  vec4 normal = vec4(a_normal, 0.0f);
+  vec4 localNorm1 = u_jointMatrices[int(a_jointIndices.x)] * normal;
+  vec4 localNorm2 = u_jointMatrices[int(a_jointIndices.y)] * normal;
+  vec4 localNorm3 = u_jointMatrices[int(a_jointIndices.z)] * normal;
+  vec4 localNorm4 = u_jointMatrices[int(a_jointIndices.w)] * normal;
+  vec4 localNorm = localNorm1 * a_jointWeights.x
+                 + localNorm2 * a_jointWeights.y
+                 + localNorm3 * a_jointWeights.z
+                 + localNorm4 * a_jointWeights.w;
+  normal = normalize(u_model * localNorm);
+
+  v_color = calculate_light(normal, u_lights, u_material);
+}
+)SHADER";
+
+const char* sFragmentShader = R"SHADER(
+precision mediump float;
+
+varying vec4 v_color;
+
+void main() {
+  gl_FragColor = vec4(v_color.xyz, 1.0f);
+}
+)SHADER";
+
+}
+
+namespace crow {
+
+struct HandGLState {
+  uint32_t jointCount;
+  std::vector<vrb::Matrix> bindMatrices;
+
+  GLuint vertexCount;
+  GLuint vboPosition;
+  GLuint vboNormal;
+  GLuint vboJointIndices;
+  GLuint vboJointWeights;
+
+  GLuint indexCount;
+  GLuint iboIndices;
+};
+
+struct HandModels::State : public vrb::ResourceGL::State {
+  GLuint vertexShader;
+  GLuint fragmentShader;
+  GLuint program;
+  GLint aPosition;
+  GLint aNormal;
+  GLint aJointIndices;
+  GLint aJointWeights;
+  GLint uJointMatrices;
+  GLint uPerspective;
+  GLint uView;
+  GLint uModel;
+  std::vector<struct HandGLState> handGLState;
+  State()
+      : vertexShader(0)
+      , fragmentShader(0)
+      , program(0)
+      , aPosition(0)
+      , aNormal(0)
+      , aJointIndices(0)
+      , aJointWeights(0)
+      , handGLState({})
+  {}
```

Comment:
Nitpick, I do prefer to initialize the attributes in the declaration
```
GLuint vertexShader { 0 };
GLuing fragmentShader { 0 };
...
````
And then you can just use the default constructor.

Reply:
Ok